### PR TITLE
Refactor Benchmarks for Less Wasm Memory Usage

### PR DESCRIPTION
**Pull request by shawntabrizi**
This extracts the `repeat` and `steps` loop of the benchmarking pipeline outside of the wasm instance, and into the client, which reduces the wasm memory needs of the benchmarking pipeline by collecting the results of each repeat using the client memory instead.

To do this, we made the following changes:

* Expose the list of available benchmarks from the runtime from the runtime apis
    * This allowed us to create a proper `--list` flag which shows the available benchmarks for the runtime
* Make it so that `dispatch_benchmark` does at most 1 benchmark (one pallet, one extrinsic, one repeat, one step) per call
    * Extracted wildcard flags for `pallet` / `extrinsic` and determining which benchmarks need to run into CLI
    * Move `steps` and `repeat` loop into the CLI
        * Modified the `steps` and `repeat` input into the runtime to be a tuple, such that we describe to the benchmark which loop and step we are on out of the total loops and steps
* Some code which re-collects all the benchmarks that are run so they are organized in an expected way for analysis 

fixes: https://github.com/paritytech/substrate/issues/9285

### CODE REVIEW

File: frame/benchmarking/src/utils.rs [hunk 1]
```diff
@@ -103,22 +103,35 @@ pub struct BenchmarkConfig {
...
+	pub steps: (u32, u32),
+	/// The number times to repeat each benchmark to increase accuracy of results. (current_repeat, total_repeat)
+	pub repeat: (u32, u32),
 	/// Enable an extra benchmark iteration which runs the verification logic for a benchmark.
 	pub verify: bool,
 	/// Enable benchmarking of "extra" extrinsics, i.e. those that are not directly used in a pallet.
 	pub extra: bool,
 }
 
+#[derive(Encode, Decode, Default, Clone, PartialEq, Debug)]
```
Comment (emostov) [inline frame/benchmarking/src/utils.rs:116]: a little doc comment would be nice

**Review by kianenigma — APPROVED** [review]

**Review by emostov — APPROVED** [review]

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -41,7 +41,7 @@ variables:                         &default-vars
   CI_IMAGE:                        "paritytech/ci-linux:production"
   # FIXME set to release
   CARGO_UNLEASH_INSTALL_PARAMS:    "--version 1.0.0-alpha.12"
-  CARGO_UNLEASH_PKG_DEF:           "--skip node node-* pallet-template pallet-example pallet-example-* subkey chain-spec-builder"
+  CARGO_UNLEASH_PKG_DEF:           "--skip node node-* pallet-template pallet-example pallet-example-* subkey chain-spec-builder frame-metadata"
```
Comment (shawntabrizi) [inline .gitlab-ci.yml:44]:
```suggestion
  CARGO_UNLEASH_PKG_DEF:           "--skip node node-* pallet-template pallet-example pallet-example-* subkey chain-spec-builder"
```

**Comment by shawntabrizi** [comment]
bot merge

**Comment by arkpar** [comment]
This was bot-merged even though it breaks polkadot build. The companion is still waiting for review.
cc @TriplEight

**Comment by TriplEight** [comment]
this means that on the next version bump we [will](https://github.com/paritytech/substrate/blob/master/.gitlab-ci.yml#L678) release the broken packages, I remember mentioning to comment out this job before this gets fixed.
It's not cool to breaking CI every time you can't merge.

**Pull request merged**